process/smartContract/process.go

```diff
@@ -1189,20 +1273,20 @@ func (sc *scProcessor) updateSmartContractCode(
 		return
 	}
 
-	transactionSender := tx.GetSndAddr()
 	currentOwner := stateAccount.GetOwnerAddress()
-	isSenderOwner := bytes.Equal(currentOwner, transactionSender)
+	isCodeDeployerOwner := bytes.Equal(currentOwner, outputAccount.CodeDeployerAddress)
 
 	noExistingCode := len(stateAccount.GetCode()) == 0
 	noExistingOwner := len(currentOwner) == 0
 	currentCodeMetadata := vmcommon.CodeMetadataFromBytes(stateAccount.GetCodeMetadata())
 	newCodeMetadata := vmcommon.CodeMetadataFromBytes(outputAccount.CodeMetadata)
 	isUpgradeable := currentCodeMetadata.Upgradeable
 	isDeployment := noExistingCode && noExistingOwner
-	isUpgrade := !isDeployment && isSenderOwner && isUpgradeable
+	isUpgrade := !isDeployment && isCodeDeployerOwner && isUpgradeable
 
 	if isDeployment {
-		stateAccount.SetOwnerAddress(transactionSender)
+		// At this point, we are under the condition "noExistingOwner"
+		stateAccount.SetOwnerAddress(outputAccount.CodeDeployerAddress)
```

Comment:
Thinking about this again - I think we should add an extra check here, to assert that `outputAccount.CodeDeployerAddress` isn't nil (that is, to protect against eventual programming errors in Arwen).

Reply:
Fixed.